rasa/nlu/tokenizers/__init__.py

```diff
@@ -14,3 +18,13 @@ def set(self, prop, info):
 
     def get(self, prop, default=None):
         return self.data.get(prop, default)
+
+    def __eq__(self, other):
```

Comment:
You're right, we need to implement ordering for `Token` or else the hashing/equals of `Message` won't work in some tests.

I think it would make sense if we also considered the `offset` and `end` properties of the `Token`:
```python
@functools.total_ordering
class Token:
    ...

    def __eq__(self, other):
        if not isinstance(other, Token):
            return NotImplemented
        return (self.offset, self.end, self.text) == (
            other.offset,
            other.end,
            other.text,
        )

    def __lt__(self, other):
        if not isinstance(other, Token):
            return NotImplemented
        return (self.offset, self.end, self.text) < (
            other.offset,
            other.end,
            other.text,
        )
```
Also note I've replaced the `raise NotImplementedError` for `return NotImplemented`. What do you think?

Reply:
I think it makes sense for the comparison to take into account `offset` and `end`. I'm not so sure about `return NotImplemented` though. Feels kinda weird doing `Token("hello", ) == 1` and not throw an error. Anyway, I've updated the unit test with this implementation.

Reply:
Not to worry, the Python docs mention:
>  `NotImplemented`: Special value which should be returned by the binary special methods (e.g. `__eq__()`, `__lt__()`, `__add__()`, `__rsub__()`, etc.) to indicate that the operation is not implemented with respect to the other type (...) [(source)](https://docs.python.org/3.6/library/constants.html#NotImplemented)

Related as well:
> A rich comparison method (e.g. `__eq__`) may return the singleton NotImplemented if it does not implement the operation for a given pair of arguments. [(source)](https://docs.python.org/3/reference/datamodel.html#object.__ge__)

Also, the docs for `functools.total_ordering` mention:
> Changed in version 3.4: Returning NotImplemented from the underlying comparison function for unrecognised types is now supported. [(source)](https://docs.python.org/3/library/functools.html#functools.total_ordering)

The names `NotImplementedError` and `NotImplemented` are confusing, as their use cases are different. It looks like `NotImplementedError` is designed to be used in user-defined methods which have not been implemented yet.